File: research-service/src/routes/research.ts
import { Router } from 'express';
import type { Logger } from 'pino';
import { z } from 'zod';
import { runResearch } from '../services/runResearch';
import { memoryStore } from '../storage/memoryStore';
import { crmClientFactory } from '../thirdparty/crmClient';
import { PlanJSON, ResearchBrief } from '../types/types';
import { scheduleSubscription } from '../subscriptions/subscriptions';

const RunSchema = z.object({
  query: z.string().min(3),
  scope: z.array(z.string()).optional(),
  geo: z.array(z.string()).optional(),
  freshness: z.union([z.string(), z.object({ months: z.number().int().positive() })]).optional(),
});

const SaveSchema = z.object({
  briefId: z.string().uuid(),
  crmRef: z
    .object({ leadId: z.string().uuid().optional(), taskId: z.string().uuid().optional() })
    .partial()
    .optional(),
  pmsRef: z
    .object({ projectId: z.string().optional(), pageId: z.string().optional() })
    .partial()
    .optional(),
  attachments: z
    .array(
      z.object({
        type: z.enum(['pdf', 'screenshot', 'other']).default('other'),
        url: z.string().url(),
        name: z.string().optional(),
      }),
    )
    .optional(),
});

const IdeasToPlanSchema = z.object({
  briefId: z.string().uuid(),
  selections: z.array(z.number().int().nonnegative()).min(1),
});

const SubscribeSchema = z.object({
  query: z.string().min(3),
  cadence: z.enum(['weekly', 'monthly']),
  geo: z.array(z.string()).optional(),
  scope: z.array(z.string()).optional(),
});

export const researchRouter = (logger: Logger) => {
  const r = Router();

  r.post('/run', async (req, res) => {
    const parsed = RunSchema.safeParse(req.body);
    if (!parsed.success) {
      return res.status(400).json({ error: parsed.error.flatten() });
    }
    try {
      const brief: ResearchBrief = await runResearch(parsed.data, logger);
      memoryStore.saveBrief(brief);
      return res.json(brief);
    } catch (err: any) {
      logger.error({ err }, 'failed to run research');
      return res.status(500).json({ error: 'research_failed' });
    }
  });

  r.post('/save', async (req, res) => {
    const parsed = SaveSchema.safeParse(req.body);
    if (!parsed.success) {
      return res.status(400).json({ error: parsed.error.flatten() });
    }
    const { briefId, crmRef, attachments } = parsed.data;
    const brief = memoryStore.getBrief(briefId);
    if (!brief) {
      return res.status(404).json({ error: 'brief_not_found' });
    }
    try {
      const crm = crmClientFactory({
        baseUrl: process.env.CRM_BASE_URL || 'https://stage-api.simpo.ai/crm',
        bearerToken: process.env.CRM_BEARER_TOKEN,
      });

      let notesId: string | undefined;
      if (crmRef) {
        const noteResp = await crm.createNote({
          subject: `Research Brief: ${brief.query} (${brief.date})`,
          description: JSON.stringify({ briefId: brief.briefId, summary: brief.summary }),
          leadId: crmRef.leadId,
        });
        notesId = noteResp.notesId;
        if (crmRef.taskId && attachments && attachments.length > 0) {
          await crm.createTaskAttachment(crmRef.taskId, {
            title: 'Research artifacts',
            description: `Artifacts for brief ${brief.briefId}`,
            attachmentUrl: attachments.map(a => ({ url: a.url, name: a.name })),
          });
        }
      }

      return res.json({ saved: true, crmNotesId: notesId });
    } catch (err: any) {
      logger.error({ err }, 'failed to save brief');
      return res.status(500).json({ error: 'save_failed' });
    }
  });

  r.post('/ideas-to-plan', async (req, res) => {
    const parsed = IdeasToPlanSchema.safeParse(req.body);
    if (!parsed.success) {
      return res.status(400).json({ error: parsed.error.flatten() });
    }
    const brief = memoryStore.getBrief(parsed.data.briefId);
    if (!brief) {
      return res.status(404).json({ error: 'brief_not_found' });
    }
    const selectedIdeas = parsed.data.selections
      .map(i => brief.ideas[i])
      .filter(Boolean);
    const plan: PlanJSON = {
      okrs: [
        {
          objective: `Execute research-driven initiatives for: ${brief.query}`,
          keyResults: [
            {
              metric: 'validated-ideas',
              target: selectedIdeas.length,
            },
          ],
        },
      ],
      initiatives: selectedIdeas.map((idea, idx) => ({
        id: `init-${idx + 1}`,
        title: idea.idea,
        tasks: [
          { id: `task-${idx + 1}-1`, title: 'Define success metrics', assignee: 'TBD' },
          { id: `task-${idx + 1}-2`, title: 'Run pilot/experiment', assignee: 'TBD' },
          { id: `task-${idx + 1}-3`, title: 'Analyze results', assignee: 'TBD' },
        ],
      })),
    };
    return res.json(plan);
  });

  r.post('/subscribe', async (req, res) => {
    const parsed = SubscribeSchema.safeParse(req.body);
    if (!parsed.success) {
      return res.status(400).json({ error: parsed.error.flatten() });
    }
    const sub = scheduleSubscription(parsed.data, logger);
    return res.json(sub);
  });

  return r;
};

File: research-service-py/app/routers/research.py
from __future__ import annotations
from fastapi import APIRouter, HTTPException
from uuid import uuid4
from ..models import (
	RunRequest,
	SaveRequest,
	IdeasToPlanRequest,
	SubscribeRequest,
	ResearchBrief,
	PlanJSON,
	OKR,
	KR,
	Initiative,
	PlanTask,
)
from ..storage.memory_store import store
from ..pipeline.run import run_research
from ..integrations.crm import CRMClient
from ..subscriptions.scheduler import schedule_subscription

router = APIRouter()


@router.post("/run", response_model=ResearchBrief)
async def run(req: RunRequest) -> ResearchBrief:
	brief = await run_research(req)
	store.save_brief(brief)
	return brief


@router.post("/save")
async def save(req: SaveRequest):
	brief = store.get_brief(req.briefId)
	if not brief:
		raise HTTPException(status_code=404, detail="brief_not_found")
	client = CRMClient()
	notes_id = None
	if req.crmRef is not None:
		notes_id = await client.create_note(
			subject=f"Research Brief: {brief.query} ({brief.date})",
			description=f"{{\"briefId\": \"{brief.briefId}\", \"summary\": \"{(brief.summary or '')[:200]}\"}}",
			lead_id=req.crmRef.get("leadId") if isinstance(req.crmRef, dict) else None,
		)
		if isinstance(req.crmRef, dict) and req.crmRef.get("taskId") and req.attachments:
			await client.create_task_attachment(
				task_id=req.crmRef["taskId"],
				title="Research artifacts",
				description=f"Artifacts for brief {brief.briefId}",
				attachment_urls=[{"url": a.url, "name": a.name or "artifact"} for a in req.attachments],
			)
	return {"saved": True, "crmNotesId": notes_id}


@router.post("/ideas-to-plan", response_model=PlanJSON)
async def ideas_to_plan(req: IdeasToPlanRequest) -> PlanJSON:
	brief = store.get_brief(req.briefId)
	if not brief:
		raise HTTPException(status_code=404, detail="brief_not_found")
	selected = [brief.ideas[i] for i in req.selections if 0 <= i < len(brief.ideas)]
	okrs = [OKR(objective=f"Execute research initiatives for: {brief.query}", keyResults=[KR(metric="validated-ideas", target=len(selected))])]
	initiatives = []
	for idx, idea in enumerate(selected, start=1):
		initiatives.append(
			Initiative(
				id=f"init-{idx}",
				title=idea.idea,
				tasks=[
					PlanTask(id=f"task-{idx}-1", title="Define success metrics", assignee="TBD"),
					PlanTask(id=f"task-{idx}-2", title="Run pilot/experiment", assignee="TBD"),
					PlanTask(id=f"task-{idx}-3", title="Analyze results", assignee="TBD"),
				],
			)
		)
	return PlanJSON(okrs=okrs, initiatives=initiatives)


@router.post("/subscribe")
async def subscribe(req: SubscribeRequest):
	sub = schedule_subscription(query=req.query, cadence=req.cadence)
	return sub